From 9f4c178d56e2b3218cd7f71be7872dd9fecf67dd Mon Sep 17 00:00:00 2001
From: huying <huying21@huawei.com>
Date: Mon, 26 Jan 2026 20:27:58 +0800
Subject: [PATCH] ums: prevent Send WR from being posted to Jetty when
 ubcore_bind_jetty fails

(cherry picked from commit 58483fd0b860b8df2784fc655996ff1d6fe39dc7)
---
 src/usock/ums/kmod/ums/cm/ums_core.h | 23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/src/usock/ums/kmod/ums/cm/ums_core.h b/src/usock/ums/kmod/ums/cm/ums_core.h
index 50644be..9538a93 100644
--- a/src/usock/ums/kmod/ums/cm/ums_core.h
+++ b/src/usock/ums/kmod/ums/cm/ums_core.h
@@ -96,9 +96,9 @@ static inline bool ums_conn_lgr_valid(const struct ums_connection *conn)
 /*
  * Returns true if the specified link is usable.
  *
- * usable means the link is ready to receive UB messages, map memory
- * on the link, etc. This doesn't ensure we are able to send UB messages
- * on this link, if sending UB messages is needed, use ums_link_sendable()
+ * A link is usable, meaning that the link has been initialized, but this does not necessarily
+ * mean that the link is ready to receive UB messages or can be used to send UB messages.
+ * If sending UB messages is needed, use ums_link_sendable().
  */
 static inline bool ums_link_usable(const struct ums_link *lnk)
 {
@@ -111,15 +111,20 @@ static inline bool ums_link_usable(const struct ums_link *lnk)
  * Returns true if the specified link is ready to receive AND send UB
  * messages.
  *
- * For the client side in first contact, the underlying QP may still in
- * RESET or RTR when the link state is ACTIVATING, checks in ums_link_usable()
- * is not strong enough. For those places that need to send any CDC or LLC
- * messages, use ums_link_sendable(), otherwise, use ums_link_usable() instead
+ * The underlying Jetty may still in RESET state when the link state is ACTIVATING,
+ * checks in ums_link_usable() is not strong enough. For those places that need to
+ * send any UB messages(e.g., Data, CDC or LLC), use ums_link_sendable(), otherwise,
+ * use ums_link_usable() instead.
  */
 static inline bool ums_link_sendable(struct ums_link *lnk)
 {
-	/* In use, the link is RTS/RTR once the jetty is bounded */
-	return ums_link_usable(lnk) && lnk->ub_jetty != NULL && lnk->ub_tjetty != NULL;
+	/*
+	 * Currently, the Jetty state transition implementation may not be consistent with
+	 * the UB protocol. Therefore, you can only confirm whether Jetty is ready to send
+	 * by checking the Jetty pointers.
+	 */
+	return ums_link_usable(lnk) && lnk->ub_jetty != NULL && lnk->ub_tjetty != NULL &&
+		lnk->ub_jetty->remote_jetty != NULL;
 }
 
 static inline bool ums_link_active(const struct ums_link *lnk)
-- 
2.28.0.windows.1

